File: scripts/fix_remaining_scrapers.js
#!/usr/bin/env node

/**
 * Bu script, kalan 9 scraper dosyasını otomatik olarak düzeltir
 * Eski upsert kodunu yeni ID-bazlı slug sistemi ile değiştirir
 */

const fs = require('fs');
const path = require('path');

const filesToFix = [
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/akbank/free.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/chippin/chippin.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/denizbank/denizbonus.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/garanti/bonus.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/vakifbank/world.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/adios.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/crystal.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/play.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/world.ts',
    '/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/ziraat/bankkart.ts',
];

// Eski pattern (regex ile bulacağız)
const oldPattern = /(\s+)(const { error } = await supabase\s+\.from\('campaigns'\)\s+\.upsert\(campaignData, { onConflict: 'reference_url' }\);)\s+(if \(error\) {[\s\S]*?} else {[\s\S]*?})/gm;

// Yeni kod (ID-bazlı sistem)
const newCode = `$1// ID-BASED SLUG SYSTEM
$1const { data: existing } = await supabase
$1    .from('campaigns')
$1    .select('id')
$1    .eq('reference_url', fullUrl || url)
$1    .single();

$1if (existing) {
$1    const finalSlug = generateCampaignSlug(title, existing.id);
$1    const { error } = await supabase
$1        .from('campaigns')
$1        .update({ ...campaignData, slug: finalSlug })
$1        .eq('id', existing.id);
$1    if (error) {
$1        console.error(\`      ❌ Update Error: \${error.message}\`);
$1    } else {
$1        console.log(\`      ✅ Updated: \${title.substring(0, 30)}... (\${finalSlug})\`);
$1    }
$1} else {
$1    const { data: inserted, error: insertError } = await supabase
$1        .from('campaigns')
$1        .insert(campaignData)
$1        .select('id')
$1        .single();
$1    if (insertError) {
$1        console.error(\`      ❌ Insert Error: \${insertError.message}\`);
$1    } else if (inserted) {
$1        const finalSlug = generateCampaignSlug(title, inserted.id);
$1        await supabase
$1            .from('campaigns')
$1            .update({ slug: finalSlug })
$1            .eq('id', inserted.id);
$1        console.log(\`      ✅ Inserted: \${title.substring(0, 30)}... (\${finalSlug})\`);
$1    }
$1}`;

let totalFixed = 0;
let totalErrors = 0;

console.log('🔧 Scraper Düzeltme Script\'i Başlatıldı...\n');

for (const filePath of filesToFix) {
    try {
        console.log(`📝 İşleniyor: ${path.basename(filePath)}`);

        if (!fs.existsSync(filePath)) {
            console.log(`   ⚠️  Dosya bulunamadı, atlanıyor...\n`);
            totalErrors++;
            continue;
        }

        let content = fs.readFileSync(filePath, 'utf8');

        // Önce eski pattern'i bulalım
        const matches = content.match(oldPattern);

        if (!matches || matches.length === 0) {
            console.log(`   ⚠️  Eski pattern bulunamadı, atlanıyor...\n`);
            totalErrors++;
            continue;
        }

        // Replace işlemi
        const newContent = content.replace(oldPattern, newCode);

        if (newContent === content) {
            console.log(`   ⚠️  Değişiklik yapılamadı, atlanıyor...\n`);
            totalErrors++;
            continue;
        }

        // Dosyayı kaydet
        fs.writeFileSync(filePath, newContent, 'utf8');
        console.log(`   ✅ Başarıyla düzeltildi!\n`);
        totalFixed++;

    } catch (error) {
        console.error(`   ❌ Hata: ${error.message}\n`);
        totalErrors++;
    }
}

console.log('═'.repeat(50));
console.log(`✅ Tamamlandı!`);
console.log(`   Düzeltilen: ${totalFixed} dosya`);
console.log(`   Hata: ${totalErrors} dosya`);
console.log('═'.repeat(50));

File: scripts/fix_remaining_scrapers.py
#!/usr/bin/env python3
"""
Kalan 6 scraper dosyasını otomatik olarak düzeltir.
Eski upsert kodunu yeni ID-bazlı slug sistemi ile değiştirir.
"""

import re

files_to_fix = [
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/play.ts",
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/yapikredi/world.ts",
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/garanti/bonus.ts",
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/denizbank/denizbonus.ts",
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/vakifbank/world.ts",
    "/Users/hipoglisemi/Desktop/kartavantaj-scraper/src/scrapers/ziraat/bankkart.ts",
]

# Eski pattern (multi-line regex)
old_pattern = r"""(\s+)const { error } = await supabase
\s+\.from\('campaigns'\)
\s+\.upsert\(campaignData, { onConflict: 'reference_url' }\);

\s+if \(error\) {
\s+console\.error\(`\s+❌ (?:Supabase Error|Error|DB Error for "[^"]*"): \$\{error(?:\?\.message \|\| 'Unknown error'|\.message)\}`\);
\s+} else {
\s+console\.log\(`\s+✅ (?:Saved|Successfully saved/updated): \$\{(?:campaignData\.title|title(?:\.substring\(0, 30\))?\.\.\.)\}`\);
\s+}"""

# Yeni kod
new_code_template = """{indent}// ID-BASED SLUG SYSTEM
{indent}const {{ data: existing }} = await supabase
{indent}    .from('campaigns')
{indent}    .select('id')
{indent}    .eq('reference_url', fullUrl)
{indent}    .single();

{indent}if (existing) {{
{indent}    const finalSlug = generateCampaignSlug(title, existing.id);
{indent}    const {{ error }} = await supabase
{indent}        .from('campaigns')
{indent}        .update({{ ...campaignData, slug: finalSlug }})
{indent}        .eq('id', existing.id);
{indent}    if (error) {{
{indent}        console.error(`      ❌ Update Error: ${{error.message}}`);
{indent}    }} else {{
{indent}        console.log(`      ✅ Updated: ${{title.substring(0, 30)}}... (${{finalSlug}})`);
{indent}    }}
{indent}}} else {{
{indent}    const {{ data: inserted, error: insertError }} = await supabase
{indent}        .from('campaigns')
{indent}        .insert(campaignData)
{indent}        .select('id')
{indent}        .single();
{indent}    if (insertError) {{
{indent}        console.error(`      ❌ Insert Error: ${{insertError.message}}`);
{indent}    }} else if (inserted) {{
{indent}        const finalSlug = generateCampaignSlug(title, inserted.id);
{indent}        await supabase
{indent}            .from('campaigns')
{indent}            .update({{ slug: finalSlug }})
{indent}            .eq('id', inserted.id);
{indent}        console.log(`      ✅ Inserted: ${{title.substring(0, 30)}}... (${{finalSlug}})`);
{indent}    }}
{indent}}}"""

total_fixed = 0
total_errors = 0

print("🔧 Scraper Düzeltme Script'i Başlatıldı...\n")

for file_path in files_to_fix:
    try:
        print(f"📝 İşleniyor: {file_path.split('/')[-1]}")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Find the old pattern
        match = re.search(old_pattern, content, re.MULTILINE)
        
        if not match:
            print(f"   ⚠️  Eski pattern bulunamadı, atlanıyor...\n")
            total_errors += 1
            continue
        
        # Get indentation from the match
        indent = match.group(1)
        
        # Generate new code with correct indentation
        new_code = new_code_template.format(indent=indent)
        
        # Replace
        new_content = re.sub(old_pattern, new_code, content, count=1, flags=re.MULTILINE)
        
        if new_content == content:
            print(f"   ⚠️  Değişiklik yapılamadı, atlanıyor...\n")
            total_errors += 1
            continue
        
        # Write back
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        
        print(f"   ✅ Başarıyla düzeltildi!\n")
        total_fixed += 1
        
    except Exception as e:
        print(f"   ❌ Hata: {e}\n")
        total_errors += 1

print("=" * 50)
print(f"✅ Tamamlandı!")
print(f"   Düzeltilen: {total_fixed} dosya")
print(f"   Hata: {total_errors} dosya")
print("=" * 50)
